src/monodroid/jni/strings.hh

```diff
@@ -0,0 +1,772 @@
+#ifndef __STRINGS_HH
+#define __STRINGS_HH
+
+#include <array>
+#include <cstring>
+#include <cerrno>
+#include <limits>
+#include <type_traits>
+#include <unistd.h>
+
+#include "platform-compat.hh"
+#include "logger.hh"
+#include "helpers.hh"
+
+namespace xamarin::android::internal
+{
+	constexpr size_t SENSIBLE_TYPE_NAME_LENGTH = 128;
+	constexpr size_t SENSIBLE_PATH_MAX = 256;
+	// 64-bit unsigned or 64-bit signed with sign
+	constexpr size_t MAX_INTEGER_DIGIT_COUNT_BASE10 = 21;
+
+#if DEBUG
+	constexpr bool BoundsCheck = true;
+#else
+	constexpr bool BoundsCheck = false;
+#endif
+
+	class string_segment
+	{
+	public:
+		force_inline bool initialized () const noexcept
+		{
+			return !_fresh;
+		}
+
+		force_inline const char* start () const noexcept
+		{
+			return _start;
+		}
+
+		force_inline size_t length () const noexcept
+		{
+			return _length;
+		}
+
+		force_inline bool empty () const noexcept
+		{
+			return length () == 0;
+		}
+
+		force_inline bool equal (const char *s) const noexcept
+		{
+			if (s == nullptr)
```

Comment:
Is it possible for `_start` to ever be `nullptr`? If so, how should such an instance compare to `segment.equal(nullptr)`?

Reply:
`_start` is `nullptr` initially and if `next_token` returns `false` on the **first** invocation. If `next_token` returns `false` then `token` should be effectively ignored - so comparison against `nullptr` passed to `equal` should always be `false` for a **valid** `string_segment` (i.e. when `next_token` returned `true`).  The `equal(nullptr, 0)` case is a bug.